windows-gaming-driver/src/setup/usb.rs

```diff
@@ -82,18 +84,116 @@ pub fn select(machine: &mut MachineConfig) -> bool {
+            } else {
+                // device not connected, handle cases of by-id and by-port
+                match binding {
+                    UsbBinding::ById(DeviceId { vendor, product }) => {
+                        let (vendor_name, product_name) = hwid::hwid_resolve_usb(vendor, product).ok().unwrap_or((None, None));
+                        let name = to_name(vendor_name, product_name);
+                        // TODO: Find out HidKind from vendor and product
+                        println!("[{}]\t{} [{:04x}:{:04x}] (Not Connected)", i, name, vendor, product);
+                    },
+                    UsbBinding::ByPort(DevicePort { bus, port }) => {
+                        println!("[{}]\tBus {} Port {} (Not Connected)", i, bus, port);
+                    }
+                }
+            }
+            last = i;
+        }
+        println!("[{}]\tNone", last + 1);
+        let index = ask::numeric("Which device would you like to remove?", 0..(last + 2));
+        if index == last + 1 {
+            break;
+        }
+        usb_devices.swap_remove(index);
+    }
+    let mut mouse = None;
+    let mut mouse_idx = None;
+    let mut keyboard = None;
+    for (i, binding) in usb_devices.iter().cloned().enumerate() {
+        if let Some(idx) = devs.iter().position(|dev| UsbBinding::from(dev.id) == binding
+                || UsbBinding::from(dev.port) ==  binding) {
+            let dev = &devs[idx];
+            if dev.kind == Some(HidKind::Mouse) && mouse == None {
+                mouse = Some(binding);
+                mouse_idx = Some(i);
+            }
+            if dev.kind == Some(HidKind::Keyboard) && keyboard == None {
+                keyboard = Some(binding);
+            }
+        }
+    }
+    if let Some(ix) = mouse_idx {
+        usb_devices.swap(0, ix);
+    }
+    (mouse, keyboard)
+}
+
+fn to_name<T: AsRef<str>>(vendor_name: Option<T>, product_name: Option<T>) -> String {
+    match (vendor_name, product_name) {
+        (None, None) => "Unknown vendor Unknown product".to_string(),
+        (Some(vendor), None) => format!("{} Unknown product", vendor.as_ref()),
+        (Some(vendor), Some(product)) => format!("{} {}", vendor.as_ref(), product.as_ref()),
+        (None, Some(_)) => unreachable!()
+    }
+}
+
 fn pick(special: Option<HidKind>,
-                 blacklist: &[UsbBinding], allow_abort: bool) -> Result<Option<DeviceId>> {
+                 blacklist: &[UsbBinding], allow_abort: bool) -> Result<Option<UsbBinding>> {
+    // TODO: let user choose between id and rt binding
+    let mut devs = list_devices(special).expect("Can't read connected usb devices");
+    let mut devs: Vec<_> = devs.drain(..)
+        .filter(|dev| !blacklist.contains(&dev.id.into()) && !blacklist.contains(&dev.port.into()))
+        .map(|dev| Some(dev))
+        .collect();
+    for (i, dev) in devs.iter().enumerate() {
+        let dev = dev.as_ref().unwrap();
```

Comment:
The list definitely contains only `Some(_)` as I just created it as such. We need `Option` though to have the "abort" `None` in there.